### PR TITLE
Fix `cupyx.ndimage.spline_filter1d` for HIP

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes an issue with some `cupyx.scipy.ndimage` routines that relied in `spline_filter` for input preprocessing in HIP environments.

This issue is similar to #4504, some of the values are not being updated in memory unless thread synchronization happens.

cc @grlee77 for future reference :)

### CODE REVIEW

File: cupyx/scipy/ndimage/_spline_prefilter_core.py [hunk 1]
```diff
@@ -169,7 +169,9 @@ def _get_spline1d_code(mode, poles, n_boundary):
         for (i = 1; i < n; ++i) {{
             c[i * element_stride] += z * c[(i - 1) * element_stride];
         }}''')
-
+        code.append('''
+        __syncthreads();
+        ''')
```
Comment (grlee77) [inline cupyx/scipy/ndimage/_spline_prefilter_core.py:174]: should this have an ifdef here as in #4504?

```suggestion
        code.append('''
        #ifdef __HIP_DEVICE_COMPILE__
        __syncthreads();
        #endif
        ''')
```

Comment (emcastillo) [reply cupyx/scipy/ndimage/_spline_prefilter_core.py]: Thank you!

**Review by asi1024 — APPROVED** [review]

**Pull request merged**